### PR TITLE
update fido2 package + 0.4 release work

### PR DESCRIPTION
**Pull request by daringer**
* `start/usb_strings.py` had circular dependency
*  scope fix in `helpers.py`

### CODE REVIEW

**Comment by daringer** [comment]
checks are ok, rough testing also fine, will merge tomorrow.
@dvzrv 0.4 release will follow shortly, which should allow arch packaging...

**Comment by dvzrv** [comment]
Good stuff! Thank you! \o/

**Comment by daringer** [comment]
this is the overhaul for v0.4:
* massive cleanups and (file-lvl) restructuring of `start` & `fido2` backends
* first run through for consistent application-wide logger usage 
* various `cli.fido2` calls for latest `fido2` package fixed
* `fido2 set-pin` and `fido2 change-pin` added
* `fido2 probe` removed as command 0x70 seems not to be implemented
* `start.update` got a lot of cleanups, added verbosity, consistency stuff (not actually re-worked yet: tbd)
* `start.update` retries (up to 3 times) and sleeps (with increasing duration) at needed points during the process
* added a lot `@todo` and `@fixme` lines to be addressed for v0.5
* added `interactive_test.sh` to run a manual, step-through test for most of the functions provided for nk-fido2 + nk-start

**Comment by szszszsz** [comment]
Testing locally

File: pynitrokey/start/upgrade_by_passwd.py [hunk 1]
```diff
@@ -427,116 +465,116 @@ def start_update(regnual, gnuk, default_password, password, wait_e, keyno, verbo
...
-    logger.debug('Final device strings: {}'.format(dev_strings_upgraded))
-    logger.debug('Finishing session {}'.format(datetime.now()))
-    local_print('Log saved to: {}'.format(UPGRADE_LOG_FN))
+        local_print("",
+            "could not connect to the device - might be due to a failed update",
+            "please re-insert the device, check version using:",
+            "$ nitropy start list")
+
+    local_print(f"device can now be safely removed from the USB slot",
+                f"final device strings: {dev_strings_upgraded}")
```
Comment (szszszsz) [inline pynitrokey/start/upgrade_by_passwd.py:574]: `dev_strings_upgraded` is not meant to be visible for user

File: pynitrokey/start/upgrade_by_passwd.py [hunk 2]
```diff
@@ -158,80 +169,103 @@ def main(wait_e, keyno, passwd, data_regnual, data_upgrade, skip_bootloader, ver
...
         mem_info = gnuk.mem_info()
+        # @todo: use global verbosity
         if verbosity:
             local_print("%08x:%08x" % mem_info)
 
-        local_print('*** Running update. Do NOT remove the device from the USB slot, '
-                    'until further notice.')
+        local_print("Running update!",
+                    "Do NOT remove the device from the USB slot, until further notice",
+                    "Downloading flash upgrade program...")
```
Comment (szszszsz) [inline pynitrokey/start/upgrade_by_passwd.py:181]: Asterisks `***` were meant to focus user on this specific lines, as many of them are printed out, and not all are important. Ideally the ones without this marker should be silenced completely to not confuse, and only output them in the log.

File: pynitrokey/start/upgrade_by_passwd.py [hunk 3]
```diff
@@ -427,116 +465,116 @@ def start_update(regnual, gnuk, default_password, password, wait_e, keyno, verbo
...
-            if 'No ICC present' in str(e):
+            local_print("error while running update")
+            str_factory_reset = "Please 'factory-reset' your device to " \
+                "continue (this will delete all user data from the device) " \
+                "and try again with PIN='12345678'"
+
+            if "No ICC present" in str(e):
                 kill_smartcard_services()
-                # local_print('*** Please run update tool again.')
+                local_print("retrying...")
```
Comment (szszszsz) [inline pynitrokey/start/upgrade_by_passwd.py:529]: This needs forced delay between attempts, 2 seconds or so.

File: pynitrokey/start/upgrade_by_passwd.py [hunk 4]
```diff
@@ -427,116 +465,116 @@ def start_update(regnual, gnuk, default_password, password, wait_e, keyno, verbo
...
-              '- reinsert device to the USB slot;\n'
-              '- run factory-reset on the device;\n'
-              '- close other applications, that possibly could use it (e.g. scdaemon, pcscd).\n')
-        exit(1)
+        local_critical("",
+            "Could not proceed with the update",
+            "Please execute one or all of the following and try again:",
+            "- re-insert device to the USB slot",
+            "- run factory-reset on the device",
+            "- close other applications, which could use it (e.g., scdaemon, pcscd)")
```
Comment (szszszsz) [inline pynitrokey/start/upgrade_by_passwd.py:549]: Device should never be removed during the update procedure. This should be executed only if the update has not started at all. To detect and correct the help strings.

File: pynitrokey/start/usb_strings.py [hunk 5]
```diff
@@ -42,6 +41,7 @@ def get_dict_for_device(dev: usb.Device) -> dict:
 
 
 def get_devices() -> list:
+    from pynitrokey.start.gnuk_token import gnuk_devices_by_vidpid
```
Comment (szszszsz) [inline pynitrokey/start/usb_strings.py:44]: `gnuk_devices_by_vidpid` should be extracted to a separate module

**Review by szszszsz — CHANGES_REQUESTED** [review]
Hi!
See comments for the requested changes. For some it perhaps would be better to apply in a separate PR.

**Comment by szszszsz** [comment]
Tested locally with:
- Nitrokey Start `RTM.10`;
- Nitrokey FIDO2 `2.0.0.nitrokey-42-g193cb45` (development).

on Fedora 31, `5.7.15-100.fc31.x86_64`.

File: pynitrokey/cli/program.py [hunk 6]
```diff
@@ -173,49 +56,47 @@ def bootloader(serial, firmware):
     Enter bootloader mode using `nitropy fido2 util program aux enter-bootloader` first.
     """
 
-    p = pynitrokey.client.find(serial)
+    from pynitrokey.fido2 import find
```
[the remark this answers is not in the document]
Comment (daringer) [reply pynitrokey/cli/program.py:59]: next PR

File: pynitrokey/cli/program.py [hunk 7]
```diff
@@ -11,145 +11,28 @@
...
+# def check_only(serial, firmware):
+#     """Validate currently flashed firmware, and run on success. Bootloader only."""
+#     from pynitrokey.fido2 import find
+#     p = find(serial)
+#     try:
+#         p.use_hid()
+#         p.program_file(firmware)
+#     except CtapError as e:
+#         if e.code == CtapError.ERR.INVALID_COMMAND:
+#             local_critical("Not in bootloader mode.", e)
```
[the remark this answers is not in the document]
Comment (daringer) [reply pynitrokey/cli/program.py:35]: for me it did flash a firmware, seemed redundant for me, will recheck towards next PR

File: pynitrokey/cli/program.py [hunk 8]
```diff
@@ -299,24 +125,35 @@ def reboot(serial):
     """Reboot.
 
     \b
-    This should reboot from anything (firmware, bootloader, DFU).
-    Separately, need to be able to set boot options.
+    This implementation actually only works for bootloader reboot
     """
 
     # this implementation actually only works for bootloader
```
[the remark this answers is not in the document]
Comment (daringer) [reply pynitrokey/cli/program.py:131]: changed the doc-string as my test showed it only worked while being inside the bootloader, will get into detail for next PR

File: pynitrokey/cli/fido2.py [hunk 9]
```diff
@@ -562,6 +562,8 @@ def reboot(serial, udp):
 
 
 fido2.add_command(rng)
+
+# @fixme: this one exists twice, once here, once in "util program aux"
 fido2.add_command(reboot)
 fido2.add_command(list)
```
[the remark this answers is not in the document]
Comment (daringer) [reply pynitrokey/cli/fido2.py]: sounds good, will do in next PR

File: interactive_test.sh [hunk 10]
```diff
@@ -101,6 +100,37 @@ function testfido2
...
 	run fido2 verify
 
+	make_title "get .hex firmware, gen sign-key, sign, (skipped: flash bad fw), flash good fw"
+	wget "https://github.com/Nitrokey/nitrokey-fido2-firmware/releases/download/2.0.0.nitrokey/nitrokey-fido2-firmware-2.0.0-app-to_sign.hex"
+	run fido2 util genkey test_key.pem
+	run fido2 util sign test_key.pem nitrokey-fido2-firmware-2.0.0-app-to_sign.hex output.json
+	
+	#echo "###>>>> THIS ONE WILL FAIL, EXPECTED FAIL:"
+	#run fido2 util program bootloader output.json
+	#sleep 1
```
[the remark this answers is not in the document]
Comment (daringer) [reply interactive_test.sh:110]: we should maybe hide commands not usable on regular devices, this is one of these...

File: interactive_test.sh [hunk 11]
```diff
@@ -0,0 +1,179 @@
+#!/bin/bash
+
+npy=venv/bin/nitropy
+
+function make_title
```
[the remark this answers is not in the document]
Comment (daringer) [reply interactive_test.sh:5]: yap

File: pynitrokey/fido2/__init__.py [hunk 12]
```diff
@@ -1,58 +1,45 @@
...
+    #    pynitrokey.fido2.force_udp_backend()
 
-def _UDP_InternalPlatformSwitch(funcname, *args, **kwargs):
-    if funcname == "__init__":
-        return HidOverUDP(*args, **kwargs)
-    return getattr(HidOverUDP, funcname)(*args, **kwargs)
+    from pynitrokey.fido2.client import NKFido2Client
+    from pynitrokey.exceptions import NoSoloFoundError
 
+    p = NKFido2Client()
```
[the remark this answers is not in the document]
Comment (daringer) [reply pynitrokey/fido2/__init__.py:16]: revived

File: pynitrokey/cli/start.py [hunk 13]
```diff
@@ -33,64 +39,67 @@ def start():
...
+@click.option("-f", "default_password", is_flag=True, default=False,
+    help=f"use default Admin PIN: {DEFAULT_PW3}")
+@click.option("-p", "password", help="use provided Admin PIN")
+@click.option("-e", "wait_e", default=DEFAULT_WAIT_FOR_REENUMERATION, type=int,
+    help="time to wait for device to enumerate, after regnual was executed on device")
+@click.option("-k", "keyno", default=0, type=int, help="selected key index")
+@click.option("-v", "verbose", default=0, type=int, help="verbosity level")
+@click.option("-y", "yes", default=False, is_flag=True, help="agree to everything")
+@click.option("-b", "skip_bootloader", default=False, is_flag=True,
+    help="Skip bootloader upload (e.g. when done so already)")
```
[the remark this answers is not in the document]
Comment (daringer) [reply pynitrokey/cli/start.py:99]: yeah, `isort` + `black` to be done in a separate PR

File: pynitrokey/start/upgrade_by_passwd.py (hunk 1 above)
Comment (daringer) [reply pynitrokey/start/upgrade_by_passwd.py:574]: next PR

File: pynitrokey/start/upgrade_by_passwd.py (hunk 2 above)
Comment (daringer) [reply pynitrokey/start/upgrade_by_passwd.py:181]: mainly done for some more output consistency, should decide on some "scheme" for user outputs, next PR(s)

File: pynitrokey/start/upgrade_by_passwd.py (hunk 3 above)
Comment (daringer) [reply pynitrokey/start/upgrade_by_passwd.py:529]: next PR

File: pynitrokey/start/usb_strings.py (hunk 5 above)
Comment (daringer) [reply pynitrokey/start/usb_strings.py:44]: next PR

**Review by szszszsz — APPROVED** [review]

**Comment by szszszsz** [comment]
@daringer 
Please merge it locally, so the merge commit will be signed with your GnuPG key.

**Comment by daringer** [comment]
oops, this was not even PR-ed against the master branch, okok, closing...

**Pull request closed without merging**